File: app/src/main/java/com/vegashot/vegashot/activity/ScheduleActivity.java
package com.vegashot.vegashot.activity;

import android.app.Activity;
import android.content.Context;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ListView;
import android.widget.TextView;

import com.vegashot.vegashot.R;
import com.vegashot.vegashot.http.VegasHotHttpClient;
import com.vegashot.vegashot.model.ClassSchedule;
import com.vegashot.vegashot.model.Clazz;
import com.vegashot.vegashot.model.DailySchedule;

import java.util.Vector;

import butterknife.ButterKnife;
import butterknife.Bind;

public class ScheduleActivity extends Activity {

    private static final String SCHEDULE = "schedule";
    private ClassSchedule mClassSchedule;

    @Bind(R.id.schedule_view_pager) ViewPager mScheduleViewPager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_schedule);

        ButterKnife.bind(this);

        if (null != savedInstanceState) {
            mClassSchedule = (ClassSchedule) savedInstanceState.getSerializable(SCHEDULE);
        }

        if (mClassSchedule == null) {
            new DownloadFilesTask().execute();
        } else {
            bindData(mClassSchedule);
        }
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        if (mClassSchedule != null) {
            outState.putSerializable(SCHEDULE, mClassSchedule);
        }
    }

    private class DownloadFilesTask extends AsyncTask<String, Integer, ClassSchedule> {
        protected ClassSchedule doInBackground(String... params) {
            return new VegasHotHttpClient().getClassSchedule();
        }

        protected void onProgressUpdate(Integer... progress) {
        }

        protected void onPostExecute(@Nullable ClassSchedule result) {
            mClassSchedule = result;
            bindData(result);
        }
    }

    class HeaderViewHolder {
        @Bind(R.id.header_title) TextView title;

        HeaderViewHolder(View view) {
            ButterKnife.bind(this, view);
        }
    }

    private void bindData(ClassSchedule schedule) {

        Vector<View> pages = new Vector<>();

        for (DailySchedule dailySchedule : schedule.dailySchedules) {
            ListView listView = new ListView(this);
            ViewGroup headerView = (ViewGroup) getLayoutInflater().inflate(R.layout.daily_schedule_header, listView, false);
            HeaderViewHolder holder = new HeaderViewHolder(headerView);

            holder.title.setText(dailySchedule.date);

            listView.setAdapter(new DailyScheduleAdapter(dailySchedule, ScheduleActivity.this));
            listView.addHeaderView(headerView);
            pages.add(listView);
        }

        SchedulePagerAdapter adapter = new SchedulePagerAdapter(this, pages);
        mScheduleViewPager.setAdapter(adapter);
    }

    class SchedulePagerAdapter extends PagerAdapter {

        private Context mContext;
        private Vector<View> pages;

        public SchedulePagerAdapter(Context context, Vector<View> pages) {
            this.mContext = context;
            this.pages = pages;
        }

        @Override
        public Object instantiateItem(ViewGroup container, int position) {
            View page = pages.get(position);
            container.addView(page);
            return page;
        }

        @Override
        public int getCount() {
            return pages.size();
        }

        @Override
        public boolean isViewFromObject(View view, Object object) {
            return view.equals(object);
        }

        @Override
        public void destroyItem(ViewGroup container, int position, Object object) {
            container.removeView((View) object);
        }

    }

    class DailyScheduleAdapter extends BaseAdapter {

        private final DailySchedule schedule;
        private final LayoutInflater inflater;

        public DailyScheduleAdapter(DailySchedule schedule, Activity activity) {
            this.schedule = schedule;
            this.inflater = activity.getLayoutInflater();
        }

        @Override
        public int getCount() {
            return schedule.getClasses().size();
        }

        @Override
        public Clazz getItem(int position) {
            return schedule.getClasses().get(position);
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        @Override
        public View getView(int position, View view, ViewGroup parent) {
            ViewHolder holder;
            if (view != null) {
                holder = (ViewHolder) view.getTag();
            } else {
                view = inflater.inflate(R.layout.schedule_list_row, parent, false);
                holder = new ViewHolder(view);
                view.setTag(holder);
            }

            final Clazz clazz = getItem(position);

            holder.time.setText(clazz.time);
            holder.name.setText(clazz.name);
            holder.instructor.setText(clazz.instructor);

            return view;
        }

        class ViewHolder {
            @Bind(R.id.time) TextView time;
            @Bind(R.id.name) TextView name;
            @Bind(R.id.instructor) TextView instructor;

            ViewHolder(View view) {
                ButterKnife.bind(this, view);
            }
        }
    }

}

File: app/src/main/java/com/vegashot/vegashot/activity/ScheduleActivity.kt
package com.vegashot.vegashot.activity

import android.app.Activity
import android.content.Context
import android.os.AsyncTask
import android.os.Bundle
import android.support.v4.view.PagerAdapter
import android.support.v4.view.ViewPager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ListView
import android.widget.TextView

import com.vegashot.vegashot.R
import com.vegashot.vegashot.http.VegasHotHttpClient
import com.vegashot.vegashot.model.ClassSchedule
import com.vegashot.vegashot.model.Clazz
import com.vegashot.vegashot.model.DailySchedule

import java.util.Vector

import butterknife.bindView

public class ScheduleActivity : Activity() {
    private var mClassSchedule: ClassSchedule? = null

    val mScheduleViewPager: ViewPager by bindView(R.id.schedule_view_pager)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_schedule)

        if (null != savedInstanceState) {
            mClassSchedule = savedInstanceState.getSerializable(SCHEDULE) as ClassSchedule?
        }

        if (mClassSchedule == null) {
            DownloadFilesTask().execute()
        } else {
            bindData(mClassSchedule!!)
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        if (mClassSchedule != null) {
            outState.putSerializable(SCHEDULE, mClassSchedule)
        }
    }

    private inner class DownloadFilesTask : AsyncTask<String, Int, ClassSchedule?>() {
        override fun doInBackground(vararg params: String): ClassSchedule? {
            return VegasHotHttpClient().getClassSchedule()
        }

        override fun onPostExecute(result: ClassSchedule?) {
            mClassSchedule = result

            if (result != null)
                bindData(result)
        }
    }

    private fun bindData(schedule: ClassSchedule) {

        val pages = Vector<View>()

        for (dailySchedule in schedule.dailySchedules) {
            val listView = ListView(this)
            val headerView = getLayoutInflater().inflate(R.layout.daily_schedule_header, listView, false) as ViewGroup

            val title: TextView? = headerView.findViewById(R.id.header_title) as? TextView
            title?.setText(dailySchedule.date)

            listView.setAdapter(DailyScheduleAdapter(dailySchedule, this@ScheduleActivity))
            listView.addHeaderView(headerView)
            pages.add(listView)
        }

        val adapter = SchedulePagerAdapter(this, pages)
        mScheduleViewPager.setAdapter(adapter)
    }

    inner class SchedulePagerAdapter(private val mContext: Context, private val pages: Vector<View>) : PagerAdapter() {

        override fun instantiateItem(container: ViewGroup, position: Int): Any {
            val page = pages.get(position)
            container.addView(page)
            return page
        }

        override fun getCount(): Int {
            return pages.size()
        }

        override fun isViewFromObject(view: View, `object`: Any): Boolean {
            return view == `object`
        }

        override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
            container.removeView(`object` as View)
        }

    }

    inner class DailyScheduleAdapter(private val schedule: DailySchedule, activity: Activity) : BaseAdapter() {
        private val inflater: LayoutInflater

        init {
            this.inflater = activity.getLayoutInflater()
        }

        override fun getCount(): Int {
            return schedule.classes.size()
        }

        override fun getItem(position: Int): Clazz {
            return schedule.classes.get(position)
        }

        override fun getItemId(position: Int): Long {
            return position.toLong()
        }

        override fun getView(position: Int, view: View?, parent: ViewGroup): View {
            val returnView: View
            if (view == null) {
                returnView = inflater.inflate(R.layout.schedule_list_row, parent, false)
            } else {
                returnView = view
            }

            val clazz = getItem(position)

            val time: TextView? = returnView.findViewById(R.id.time) as? TextView
            time?.setText(clazz.time)

            val name: TextView? = returnView.findViewById(R.id.name) as? TextView
            name?.setText(clazz.name)

            val instructor: TextView? = returnView.findViewById(R.id.instructor) as? TextView
            instructor?.setText(clazz.instructor)

            return returnView
        }
    }

    companion object {

        private val SCHEDULE = "schedule"
    }

}
